### PR TITLE
moved slice by class id before shape matching

### PR DESCRIPTION
**Pull request by simian-typist**
Addressing issue https://github.com/tensorflow/tensorflow/issues/30983

If you passed sample weights and a class id to a keras metric it would
enforce that the predictions, labels and weights had the same shape
before slicing the labels and predictions, resulting in a shape mismatch
when trying to apply the weights later.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by pavithrasv** [comment]
Is there a test case that you can add to verify this change?

**Comment by simian-typist** [comment]
Sure thing, I'll write one when I get home.

**Comment by simian-typist** [comment]
Whoops, added the test but pushed from the wrong account. so gonna roll it back and re-push. Sorry about that.

**Comment by simian-typist** [comment]
Not the best test I've ever written, but it'll fail if the sample weights aren't being broadcast (or rather not broadcast) correctly.

**Comment by gbaned** [comment]
@simian-typist Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Review by pavithrasv — DISMISSED** [review]

**Review by pavithrasv — DISMISSED** [review]

**Review by pavithrasv — APPROVED** [review]
